### PR TITLE
Serializable datasets

### PR DESCRIPTION
**Pull request by bartvm**
Removed the generators used in schemes and replaced them with custom
iterators so that dill can serialize them. Also added a test that
serializes an MNIST data stream and deserializes it, checking whether
the state is maintained.

@rizar

### CODE REVIEW

File: blocks/datasets/schemes.py [hunk 1]
```diff
@@ -91,5 +92,22 @@ def __init__(self, num_examples, batch_size):
         self.batch_size = batch_size
 
     def get_request_iterator(self):
-        return (slice(x, min(self.num_examples, x + self.batch_size))
-                for x in range(0, self.num_examples, self.batch_size))
+        return SequentialIterator(self.num_examples, self.batch_size)
+
+
+class SequentialIterator(six.Iterator):
```
Comment (rizar) [inline blocks/datasets/schemes.py]: Having such an iterator is also better in terms of memory, the old approach allocated a potentially quite large list.

**Comment by rizar** [comment]
Great, however I am confused that pickling `repeat` works for you, because it does not work for me:

```
In [12]: dill.dump(repeat(3), open("f.pkl", "w"))

In [13]: dill.load(open("f.pkl"))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-13-788f69d0c637> in <module>()
----> 1 dill.load(open("f.pkl"))

/usr/local/lib/python2.7/dist-packages/dill/dill.pyc in load(file)
    197     pik = Unpickler(file)
    198     pik._main_module = _main_module
--> 199     obj = pik.load()
    200     if type(obj).__module__ == _main_module.__name__: # point obj class to main
    201         try: obj.__class__ == getattr(pik._main_module, type(obj).__name__)

/usr/lib/python2.7/pickle.pyc in load(self)
    856             while 1:
    857                 key = read(1)
--> 858                 dispatch[key](self)
    859         except _Stop, stopinst:
    860             return stopinst.value

/usr/lib/python2.7/pickle.pyc in load_newobj(self)
   1081         args = self.stack.pop()
   1082         cls = self.stack[-1]
-> 1083         obj = cls.__new__(cls, *args)
   1084         self.stack[-1] = obj
   1085     dispatch[NEWOBJ] = load_newobj

TypeError: Required argument 'object' (pos 1) not found

In [14]: 
```

**Comment by rizar** [comment]
Okay, I guess I see what happened. It worked for me in Python 3, and perhaps you tried locally in Python 3 and your MNIST test uses the `SequentialScheme` only. If that's the case, can you please reintroduce your `ConstantIterator`?

File: blocks/datasets/schemes.py (hunk 1 above)
Comment (bartvm) [inline blocks/datasets/schemes.py]: No, it was never a list, the parentheses made it a generator, so it used just as much memory.

Comment (rizar) [inline blocks/datasets/schemes.py]: My bad, I thought it was a big tuple.

**Pull request merged**